Bump cloud-provider-gcp to v1.21

cici37 opened the pull request:
This PR will:
- Updating cloud-provider-gcp to both deploy with and build against K/K v1.20.
- Updated the libraries we build against.
- Update //providers/gce to the latest v1.21(v1.21.1-rc.0)

Summary of the changes:
- Changes under /vendor are auto generated
- Update release hash in WORKSPACE
- Update image tagging under /tools
- Update go.mod with v1.21 release
- Ping version on go.mod
- Files under //cmd/cloud-controll-manager are due to flag passing fix in upstream
- Have separate go module under //providers
- Update //provider/go.mod with v1.21 release
- Update files under //provider/gec to the latest v1.21(v1.21.1-rc.0)
- Explicitly import k8s.io/cloud-provider/credentialconfig to handle [issue](https://github.com/kubernetes/cloud-provider-gcp/issues/211)
- Added serviceaccounts/token permission to system:cloud-controller-manager
- Update config-test.sh to fix conformance test failure

cici37 commented:
Wait for #207 to be merged first

jiahuif commented:
To ease reviewing, could you split the change into two commits? First for human-made changes, second for any changes made by tools.

/assign

cici37 commented:
/retest cloud-provider-gcp-e2e-create

cici37 commented:
/test cloud-provider-gcp-e2e-create

cici37 commented:
Summary of the changes:
- Changes under /vendor are auto generated
- Update release hash in WORKSPACE
- Update image tagging under /tools
- Update go.mod with v1.21 release
- Ping version on go.mod
- Files under //cmd/cloud-controll-manager are due to flag passing fix in upstream
- Have separate go module under //providers
- Update //provider/go.mod with v1.21 release
- Update files under //provider/gec to the latest v1.21(v1.21.1-rc.0)
- Explicitly import k8s.io/cloud-provider/credentialconfig to handle [issue](https://github.com/kubernetes/cloud-provider-gcp/issues/211)
- Added serviceaccounts/token permission to system:cloud-controller-manager
@jiahuif @jpbetz This PR is ready to review now. I was running conformance tests locally and will update the results when it is ready.

jiahuif commented on providers/gce/OWNERS:
```diff
@@ -1,6 +1,7 @@
 # See the OWNERS docs at https://go.k8s.io/owners
-
-approvers:
+# We are no longer accepting features into k8s.io/legacy-cloud-providers.
```
Maybe we do not need these statements in out-of-tree providers.

jpbetz replied:
Let's also add @jiahuif to reviewers (in addition to you and I).

cici37 replied:
Updated and added @jiahuif

jiahuif commented on tools/version.sh:
```diff
@@ -74,7 +74,7 @@ get_version_vars() {
     # Use git describe to find the version based on tags.
     #if [[ -n ${KUBE_GIT_VERSION-} ]] || KUBE_GIT_VERSION=$("${git[@]}" describe --tags --match='v*' --abbrev=14 "${KUBE_GIT_COMMIT}^{commit}" 2>/dev/null); then
-    if [[ -n ${KUBE_GIT_VERSION-} ]] || KUBE_GIT_VERSION=$(echo "v1.20.0+${KUBE_GIT_COMMIT}"); then
+    if [[ -n ${KUBE_GIT_VERSION-} ]] || KUBE_GIT_VERSION=$(echo "v1.21.0-alpha+${KUBE_GIT_COMMIT}"); then
```
Is the `alpha` part still needed as of today?

cici37 replied:
It comes from her: https://github.com/kubernetes/cloud-provider-gcp/pull/178#discussion_r599971852. And I think it would make sense to indicate the local build. What do you think?

jiahuif replied:
Let's keep the "alpha" then, just in case we have a chance to use `k/k` master

cici37 commented:
> To ease reviewing, could you split the change into two commits? First for human-made changes, second for any changes made by tools.
> 
> /assign

Update:
The commits are split into three:
- Copied files from k8s.io/legacy-cloud-providers/gce for v1.21 to //providers/gce
- Manual changes to update cloud-provider-gcp to v1.21
- Auto changes to update cloud-provider-gcp to v1.21

Thanks for the suggestion. Hope it would help with review.

jiahuif commented on cluster/addons/cloud-controller-manager/cloud-node-controller-role.yaml:
```diff
@@ -154,4 +160,4 @@ rules:
   verbs:
   - list
-  - delete
+  - delete
```
Let's be more care with these straw edits, might pollute history.

cici37 replied:
Updated

jiahuif commented on cluster/addons/cloud-controller-manager/cloud-node-controller-role.yaml:
```diff
@@ -80,6 +80,12 @@ rules:
   - list
   - watch
+- apiGroups:
+    - ""
```
The indents looks a bit off. Might need to check your editor settings.

cici37 replied:
Updated

jiahuif commented:
`node_controller` does not seem to be running correctly.

jpbetz commented:
/assign @cheftako There are a couple minor outstanding feedback items from @jiahuif and I but I think this is otherwise ready for you to have a look at.

jpbetz commented:
/assign
/assign @cheftako
There are a couple minor outstanding feedback items from @jiahuif and I but I think this is otherwise ready for you to have a look at.

cici37 commented on cluster/addons/cloud-controller-manager/cloud-node-controller-role.yaml:
```diff
@@ -154,4 +160,4 @@ rules:
   verbs:
   - list
-  - delete
+  - delete
```
Yes, Already updated

cici37 commented on providers/go.mod:
```diff
@@ -11,19 +11,29 @@ require (
-	k8s.io/client-go => k8s.io/client-go v0.20.0
-	k8s.io/cloud-provider => k8s.io/cloud-provider v0.20.0
-	k8s.io/component-base => k8s.io/component-base v0.20.0
+    cloud.google.com/go => cloud.google.com/go v0.75.0
```
Thanks for pointing out. Somehow my IDE editor didn't show the indentation difference. Updated

cici37 commented on providers/go.mod:
```diff
@@ -11,19 +11,29 @@ require (
+	k8s.io/cloud-provider => k8s.io/cloud-provider v0.21.0
+	k8s.io/component-base => k8s.io/component-base v0.21.0
+	k8s.io/klog/v2 => k8s.io/klog/v2 v2.8.0
+    k8s.io/utils => k8s.io/utils v0.0.0-20210111153108-fddb29f9d009
```
Upated

cheftako commented on cmd/cloud-controller-manager/main.go:
```diff
@@ -20,27 +20,70 @@ limitations under the License.
+	fss := cliflag.NamedFlagSets{}
+	command := app.NewCloudControllerManagerCommand(ccmOptions, cloudInitializer, app.DefaultInitFuncConstructors, fss, wait.NeverStop)
+
+	// TODO: once we switch everything over to Cobra commands, we can go back to calling
```
Can we have a bug referenced for this TODO?

cici37 replied:
It was been added by https://github.com/kubernetes/kubernetes/pull/58407. And together within some other files(like `https://github.com/kubernetes/kubernetes/blob/271476446b32bfc69a95ea20d8b02d5303128f5e/cmd/kube-proxy/proxy.go#L39`). I will create an issue for tracking.

cheftako replied:
Can we have a separate issue for this repo to fix it here. Then reference that issue in this TODO?

cici37 replied:
Updated. Ref: https://github.com/kubernetes/cloud-provider-gcp/issues/215

cheftako commented on cmd/cloud-controller-manager/main.go:
```diff
@@ -20,27 +20,70 @@ limitations under the License.
+	// initialize cloud provider with the cloud provider name and config file provided
+	cloud, err := cloudprovider.InitCloudProvider(cloudConfig.Name, cloudConfig.CloudConfigFile)
+	if err != nil {
+		klog.Fatalf("Cloud provider could not be initialized: %v", err)
```
Please add cloudConfig.Name and cloudConfig.CloudConfigFile to the outputted fatal message.

cici37 replied:
Updated

cheftako commented on cmd/cloud-controller-manager/main.go:
```diff
@@ -20,27 +20,70 @@ limitations under the License.
+		klog.Fatalf("Cloud provider could not be initialized: %v", err)
+	}
+	if cloud == nil {
+		klog.Fatalf("Cloud provider is nil")
```
Ditto adding cloudConfig.Name and cloudConfig.CloudConfigFile.

cici37 replied:
Updated

cheftako commented on providers/gce/OWNERS:
```diff
@@ -1,6 +1,5 @@
 # See the OWNERS docs at https://go.k8s.io/owners
-
-approvers:
+emeritus_approvers:
```
You seem to have removed all approvers for this directory. Do we want that?

cici37 replied:
Thanks for catching!(It's been removed from legacy-cloud-providers/gce.) I have added back for //providers/gce/OWNERS.

jpbetz commented:
/test ci-cloud-provider-gcp-conformance-latest

(to verify that kubetest2 is now passing with the log-file -> log_file fix)

cheftako commented on cmd/cloud-controller-manager/main.go:
```diff
@@ -20,27 +20,70 @@ limitations under the License.
+	"k8s.io/klog/v2"
+
+	_ "k8s.io/cloud-provider-gcp/providers/gce"
+	// For existing cloud providers, the option to import legacy providers is still available.
```
These two comment lines should be removed.

cici37 replied:
Updated

cici37 commented:
Update:
- Update config-test.sh to fix conformance test failure.
- Addressed all comments

Current PR passed all conformance tests:
![Screen Shot 2021-05-03 at 4 19 08 PM](https://user-images.githubusercontent.com/8658046/116945165-e1650f80-ac2b-11eb-8517-7c8046878e62.png)

cheftako commented on cluster/gce/manifests/cloud-controller-manager.manifest:
```diff
@@ -26,7 +26,7 @@
     "command": ["/cloud-controller-manager"],
     "args": [
-      "--log-file=/var/log/cloud-controller-manager.log",
+      "--log_file=/var/log/cloud-controller-manager.log",
```
I notice we have a smattering of manifests and the configure-helper.sh script which still set the --log-file flag for other processes. Not saying we should do anything in this PR but we should look at adding some sort of test for this. This flag is set by klog which indicates we can expect updating versions of the other binaries can cause the same problem in them.

cici37 replied:
Thanks for adding the info. Other process with go-runner will expect `--log-file`(e.g.: https://github.com/kubernetes/cloud-provider-gcp/blob/c497d0da8d551515b61fb132612718d390809814/cluster/gce/manifests/kube-controller-manager.manifest#L28).

cheftako replied:
:+1:

cheftako commented:
/lgtm
/approve
